src/stan/io/json/json_data_handler.hpp

```diff
@@ -81,100 +369,156 @@ class json_data_handler : public stan::json::json_handler {
-    if (dims_.size() < dim_idx_) {
-      dims_.push_back(0);
-      dims_unknown_.push_back(true);
-      dims_verify_.push_back(0);
-    } else {
-      dims_verify_[dim_idx_ - 1] = 0;
+  }
+
+  void start_object() {
+    event = meta_event::OBJ_OPEN;
+    if (is_init())
+      return;
+    if (slot_types_map[key_str()] == meta_type::ARRAY) {
+      slot_types_map[key_str()] = meta_type::ARRAY_OF_TUPLES;
+    } else if (slot_types_map[key_str()] == meta_type::SCALAR) {
+      slot_types_map[key_str()] = meta_type::TUPLE;
     }
+    // don't reset tuples or array of tuples
   }
 
-  void end_array() {
-    if (dims_unknown_[dim_idx_ - 1] == true) {
-      dims_unknown_[dim_idx_ - 1] = false;
-    } else if (dims_verify_[dim_idx_ - 1] != dims_[dim_idx_ - 1]) {
-      std::stringstream errorMsg;
-      errorMsg << "variable: " << key_ << ", error: non-rectangular array";
-      throw json_error(errorMsg.str());
+  void end_object() {
+    event = meta_event::OBJ_CLOSE;
+    if (key_stack.size() > 1
+        && slot_types_map[outer_key_str()] == meta_type::ARRAY_OF_TUPLES) {
+      array_dims outer = get_outer_dims(key_stack);
+      if (!outer.dims.empty()) {
+        outer.dims_acc[outer.dims.size() - 1]++;
+        set_outer_dims(outer);
+      }
     }
-    if (0 == dim_last_
-        && ((is_int_ && values_i_.size() > 0) || (values_r_.size() > 0)))
-      dim_last_ = dim_idx_;
-    dim_idx_--;
+    save_key_value_pair();
   }
 
-  void start_object() {
-    if (!is_init()) {
+  void start_array() {
+    if (key_stack.empty()) {
+      throw json_error("Expecting JSON object, found array.");
+    }
+    if (slot_types_map[key_str()] == meta_type::SCALAR
+        && !(values_r.empty() && values_r.empty())) {
       std::stringstream errorMsg;
-      errorMsg << "variable: " << key_ << ", error: nested objects not allowed";
+      errorMsg << "Variable: " << key_str()
+               << ", error: non-scalar array value.";
       throw json_error(errorMsg.str());
     }
+    if (slot_types_map[key_str()] == meta_type::SCALAR)
+      slot_types_map[key_str()] = meta_type::ARRAY;
+    else if (slot_types_map[key_str()] == meta_type::TUPLE)
+      unexpected_error(key_str());
+    array_dims dims;
+    if (slot_dims_map.count(key_str()) == 1)
+      dims = slot_dims_map[key_str()];
+    dims.cur_dim++;
+    if (dims.dims.empty() || dims.dims.size() < dims.cur_dim) {
+      dims.dims.push_back(0);
+      dims.dims_acc.push_back(0);
+    }
+    if (dims.cur_dim > 1)
+      dims.dims_acc[dims.cur_dim - 2]++;
+    slot_dims_map[key_str()] = dims;
+    array_start_i = values_i.size();
+    array_start_r = values_r.size();
   }
 
-  void end_object() {
-    save_current_key_value_pair();
-    reset();
-  }
-
-  void promote_to_double() {
-    if (is_int_) {
-      for (std::vector<int>::iterator it = values_i_.begin();
-           it != values_i_.end(); ++it)
-        values_r_.push_back(*it);
-      is_int_ = false;
+  void end_array() {
```

Comment:
I'd appreciate a couple high-level comments which provide an overview of the relationships between these functions/the invariants between them

Reply:
added high-level comments to the event handlers for key as well as tuple and array open, close.
this was a very good way to identify code cruft as well as a way to tighten up processing - in particular, for single-dimensional arrays, we don't need to do anything to the vector of values, just save it.